### ISSUE TITLE
[Bug]: illegal character encoding in string literal (chinese)

### ISSUE THREAD

**Issue by Leiyi548** [LunarVim/LunarVim#1792]
### Problem description

hai,I'm from China. I like lunarvim very much. I think it's cool. When I want to input Chinese, LSP reminds me that this is a wrong character code. How can I solve this problem?

### Check the following

- [X] I am in a git managed directory
- [X] I restarted Neovim after running :LspInstall
- [X] I have a valid language configuration (~/.config/lvim/ftplugin/)
- [X] I checked the options (~/.config/lvim/config.lua)
- [ ] The programs specified by the config are installed

### LunarVim version

latest

### Neovim version (>= 0.5)

0.5.0

### Screenshots

![image](https://user-images.githubusercontent.com/76139422/137612471-09a5e89b-5686-4ac1-b083-95d6e3b950f4.png)

### I have

- [X] Read the readme
- [X] Read the wiki
- [X] Searched for similar issues

**Comment by kylo252** [comment]
Does this only happen with C/C++ when `clangd` is running?

**Comment by kylo252** [comment]
This is sadly an upstream issue, see https://github.com/neovim/neovim/issues/15116

**Comment by Leiyi548** [comment]
@kylo252  
![image](https://user-images.githubusercontent.com/76139422/137618563-7238c465-eaf2-4969-9dce-dd9a652369ea.png)
I can't run your command. I don't have this file in this folder

**Comment by kylo252** [comment]
Sorry, I forgot to mention that you need to do `LvimUpdate`. I had just fixed that file before my comment.

**Comment by kylo252** [comment]
> In China, LvimUpdate failed. How can I update it?

It's the usual way you update lunarvim, you can download the file manually and use if from anywhere as well
https://github.com/LunarVim/LunarVim/blob/0744e9a5e9e6795ac7e3a7f2f41f89932360b57a/tests/minimal_lsp.lua

**Comment by kylo252** [comment]
I think you need to increase the log level for `clangd` as well. 
```lua
cmd = default_opts.cmd
-- add this 
vim.list_extend(cmd, {"--log=verbose", "--pretty"})
```


> ```logtalk
> [ ERROR ] 2021-10-17T22:53:24+0800 ] ...llar/neovim/0.5.0/share/nvim/runtime/lua/vim/lsp/rpc.lua:462 ]	"rpc"	"/Users/macos/.local/share/nvim/lsp_servers/clangd/clangd"	"stderr"	"I[22:53:24.776] clangd version 12.0.1 (https://github.com/llvm/llvm-project fed41342a82f5a3a9201819a82bf7a48313e296b)\n"
> [ ERROR ] 2021-10-17T22:53:24+0800 ] ...llar/neovim/0.5.0/share/nvim/runtime/lua/vim/lsp/rpc.lua:462 ]	"rpc"	"/Users/macos/.local/share/nvim/lsp_servers/clangd/clangd"	"stderr"	"I[22:53:24.776] PID: 76265\nI[22:53:24.776] Working directory: /Users/macos/Desktop/leetcode/dayPratice/C++\nI[22:53:24.776] argv[0]: 
> ```

Also make sure that you're triggering the encoding warning so that it shows up in the the log.

**Comment by mjlbach** [comment]
All we care about is rpc.send, the default log level is fine

```
vim.lsp.set_log_level("debug")
require('vim.lsp.log').set_format_func(vim.inspect)
```

**Comment by mjlbach** [comment]
https://github.com/neovim/neovim/pull/16121 is ready for testing now.

**Comment by howIsPassword** [comment]
![输入中文失败2](https://user-images.githubusercontent.com/57863647/139581246-5d3518ac-391d-4f97-a3f3-963c7237e6c2.gif)
@mjlbach There is no problem entering Chinese comments. Entering Chinese in the string will still use errors. After restarting LSP, the errors disappear.

**Comment by mjlbach** [comment]
I don't understand. Did you test the new PR?

**Comment by howIsPassword** [comment]
yes，I am builed
![image](https://user-images.githubusercontent.com/57863647/139581496-9712c370-49b2-434b-975a-d516ae0dc58e.png)

**Comment by mjlbach** [comment]
Can you send me the rpc log? if you use the minimal init.lua from lspconfig it should set everything up for you (and clear ~/.cache/nvim/lsp.log). Specifically clearing the log and only send me it during entering the string portion (not comment)

**Comment by mjlbach** [comment]
Another thing to try is:

`capabilities=vim.lsp.protocol.make_client_capabilities()` for the `setup` call to clangd to force utf-16 (on that PR)

**Comment by howIsPassword** [comment]
When I use minimal init.lua, there is nothing in the log file,
Using the default configuration, the log has something
[lsp.log](https://github.com/LunarVim/LunarVim/files/7448741/lsp.log)
![minilsp没有问题](https://user-images.githubusercontent.com/57863647/139582475-efe2c7db-bc6e-474e-a682-edc327e8e6c3.gif)
![默认配置有问题](https://user-images.githubusercontent.com/57863647/139582484-b3f700c6-0ed5-4792-ba5d-9dfe30d4be5e.gif)

**Comment by mjlbach** [comment]
Unfortunately there's a lot of noise there, you can try this:

```
vim.cmd [[set runtimepath=$VIMRUNTIME]]
vim.cmd [[set packpath=/tmp/nvim/site]]

local package_root = '/tmp/nvim/site/pack'
local install_path = package_root .. '/packer/start/packer.nvim'

local function load_plugins()
  require('packer').startup {
    {
      'wbthomason/packer.nvim',
      'neovim/nvim-lspconfig',
    },
    config = {
      package_root = package_root,
      compile_path = install_path .. '/plugin/packer_compiled.lua',
    },
  }
end

_G.load_config = function()
  vim.lsp.set_log_level 'trace'
  if vim.fn.has 'nvim-0.5.1' == 1 then
    require('vim.lsp.log').set_format_func(vim.inspect)
  end
  local nvim_lsp = require 'lspconfig'
  local on_attach = function(_, bufnr)
    local function buf_set_keymap(...)
      vim.api.nvim_buf_set_keymap(bufnr, ...)
    end
    local function buf_set_option(...)
      vim.api.nvim_buf_set_option(bufnr, ...)
    end

    buf_set_option('omnifunc', 'v:lua.vim.lsp.omnifunc')

    -- Mappings.
    local opts = { noremap = true, silent = true }
    buf_set_keymap('n', 'gD', '<Cmd>lua vim.lsp.buf.declaration()<CR>', opts)
    buf_set_keymap('n', 'gd', '<Cmd>lua vim.lsp.buf.definition()<CR>', opts)
    buf_set_keymap('n', 'K', '<Cmd>lua vim.lsp.buf.hover()<CR>', opts)
    buf_set_keymap('n', 'gi', '<cmd>lua vim.lsp.buf.implementation()<CR>', opts)
    buf_set_keymap('n', '<C-k>', '<cmd>lua vim.lsp.buf.signature_help()<CR>', opts)
    buf_set_keymap('n', '<space>wa', '<cmd>lua vim.lsp.buf.add_workspace_folder()<CR>', opts)
    buf_set_keymap('n', '<space>wr', '<cmd>lua vim.lsp.buf.remove_workspace_folder()<CR>', opts)
    buf_set_keymap('n', '<space>wl', '<cmd>lua print(vim.inspect(vim.lsp.buf.list_workspace_folders()))<CR>', opts)
    buf_set_keymap('n', '<space>D', '<cmd>lua vim.lsp.buf.type_definition()<CR>', opts)
    buf_set_keymap('n', '<space>rn', '<cmd>lua vim.lsp.buf.rename()<CR>', opts)
    buf_set_keymap('n', 'gr', '<cmd>lua vim.lsp.buf.references()<CR>', opts)
    buf_set_keymap('n', '<space>e', '<cmd>lua vim.lsp.diagnostic.show_line_diagnostics()<CR>', opts)
    buf_set_keymap('n', '[d', '<cmd>lua vim.lsp.diagnostic.goto_prev()<CR>', opts)
    buf_set_keymap('n', ']d', '<cmd>lua vim.lsp.diagnostic.goto_next()<CR>', opts)
    buf_set_keymap('n', '<space>q', '<cmd>lua vim.lsp.diagnostic.set_loclist()<CR>', opts)
  end

  nvim_lsp['clangd'].setup {
    capabilities = vim.lsp.protocol.make_client_capabilities(),
    on_attach = on_attach,
  }

  print [[You can find your log at $HOME/.cache/nvim/lsp.log. Please paste in a github issue under a details tag as described in the issue template.]]
end

if vim.fn.isdirectory(install_path) == 0 then
  vim.fn.system { 'git', 'clone', 'https://github.com/wbthomason/packer.nvim', install_path }
  load_plugins()
  require('packer').sync()
  vim.cmd [[autocmd User PackerComplete ++once lua load_config()]]
else
  load_plugins()
  require('packer').sync()
  _G.load_config()
end
```

**Comment by mjlbach** [comment]
You should get a logfile that looks like:

```
[DEBUG][2021-10-31 05:16:24] .../vim/lsp/rpc.lua:339	"rpc.send"	{
  jsonrpc = "2.0",
  method = "textDocument/didChange",
  params = {
    contentChanges = { {
        range = {
          end = {
            character = 0,
            line = 9
          },
          start = {
            character = 0,
            line = 9
          }
        },
        rangeLength = 0,
        text = "#"
      } },
    textDocument = {
      uri = "file:///home/michael/Repositories/neovim_development/nvim-lspconfig/neovim/src/nvim/main.c",
      version = 4
    }
  }
}
```

**Comment by mjlbach** [comment]
Note, I overrode capabilities in clangd to force it to use utf-16 encoding to confirm a theory, you can comment that line to get back to the "default" utf-8 native behavior.

**Comment by howIsPassword** [comment]
This is obtained by using your configuration file
[lsp.log](https://github.com/LunarVim/LunarVim/files/7448814/lsp.log)
![新配置文件](https://user-images.githubusercontent.com/57863647/139583198-63750b69-9730-4197-af62-a859e3ac9a07.gif)

**Comment by mjlbach** [comment]
Do you see an error there? I don't see anything in the log after you finish line 2

**Comment by howIsPassword** [comment]
There is no error in using the configuration file you gave, but there is a problem in using the default configuration. How can I use ` capabilities = vim. LSP. Protocol. Make in the default configuration_ client_ Capabilities () `. Is that right?
![image](https://user-images.githubusercontent.com/57863647/139584500-9172181f-ba62-4df7-8d04-26b5316ba8ad.png)

**Comment by mjlbach** [comment]
Sorry, can you do one more test for me? Can you remove that line (capabilities) and see if the issue returns? If so, I believe I know exactly what it is.

I'm not sure what lvim.lsp.manager is, I only work on neovim, and don't use lunarvim

**Comment by howIsPassword** [comment]
But when I comment on these three lines, the problem still exists
![image](https://user-images.githubusercontent.com/57863647/139585024-ffce7351-87e4-4ed5-b378-bff991b98daa.png)
![image](https://user-images.githubusercontent.com/57863647/139585036-caffc79f-a7be-4f21-92aa-63e5740e146c.png)
lvim.lsp.manager is this
```
local M = {}

local Log = require "lvim.core.log"
local lvim_lsp_utils = require "lvim.lsp.utils"

function M.init_defaults(languages)
  for _, entry in ipairs(languages) do
    if not lvim.lang[entry] then
      lvim.lang[entry] = {
        formatters = {},
        linters = {},
        lsp = {},
      }
    end
  end
end

---Resolve the configuration for a server based on both common and user configuration
---@param name string
---@param user_config table [optional]
---@return table
local function resolve_config(name, user_config)
  local config = {
    on_attach = require("lvim.lsp").common_on_attach,
    on_init = require("lvim.lsp").common_on_init,
    capabilities = require("lvim.lsp").common_capabilities(),
  }

  local has_custom_provider, custom_config = pcall(require, "lvim/lsp/providers/" .. name)
  if has_custom_provider then
    Log:debug("Using custom configuration for requested server: " .. name)
    config = vim.tbl_deep_extend("force", config, custom_config)
  end

  if user_config then
    config = vim.tbl_deep_extend("force", config, user_config)
  end

  return config
end

---Setup a language server by providing a name
---@param server_name string name of the language server
---@param user_config table [optional] when available it will take predence over any default configurations
function M.setup(server_name, user_config)
  vim.validate { name = { server_name, "string" } }

  if lvim_lsp_utils.is_client_active(server_name) then
    return
  end

  local config = resolve_config(server_name, user_config)
  local server_available, requested_server = require("nvim-lsp-installer.servers").get_server(server_name)

  local function ensure_installed(server)
    if server:is_installed() then
      return true
    end
    if not lvim.lsp.automatic_servers_installation then
      Log:debug(server.name .. " is not managed by the automatic installer")
      return false
    end
    Log:debug(string.format("Installing [%s]", server.name))
    server:install()
    vim.schedule(function()
      vim.cmd [[LspStart]]
    end)
  end

  if server_available and ensure_installed(requested_server) then
    requested_server:setup(config)
  else
    -- since it may not be installed, don't attempt to configure the LSP unless there is a custom provider
    local has_custom_provider, _ = pcall(require, "lvim/lsp/providers/" .. server_name)
    if has_custom_provider then
      require("lspconfig")[server_name].setup(config)
    end
  end
end

return M
```

**Comment by kylo252** [comment]
> But when I comment on these three lines, the problem still exists

@howIsPassword, Which one of these has the problem?
```lua
local opts = {
  capabilities = vim.lsp.make_client_capabilities()
}
require("lvim.lsp.manager").setup("clangd", opts)
```
vs

```lua
require("lvim.lsp.manager").setup("clangd")
```

---

Note: Please check that you're only setting up `clangd` once. 

```lua
vim.list_extend(lvim.lsp.override, { "clangd" }) --followed by `:LvimCacheReset`
```

**Comment by mjlbach** [comment]
I'm beginning to think this was a red herring from core side, and is actually a lunarvim issue, try *not* overriding capabilities or on_init for clangd.

**Comment by kylo252** [comment]
> I'm beginning to think this was a red herring from core side, and is actually a lunarvim issue, try _not_ overriding capabilities or on_init for clangd.

isn't this supposed to be standard? 😕 

https://github.com/LunarVim/LunarVim/blob/e8693406babee724dd51293dc6dad10931dbef45/lua/lvim/lsp/init.lua#L69-L86

if anything, it's `lspconfig` that's enforcing the encoding: https://github.com/neovim/nvim-lspconfig/blob/dd7a616b8d778c195fe7ec9fffa32ce1b0fc9da7/lua/lspconfig/clangd.lua#L36

**Comment by mjlbach** [comment]
I'm suggesting you special case clangd in lunarvim for now for the sake of debugging. I'm unable to reproduce the issue with the minimal config.

**Comment by mjlbach** [comment]
TLDR it's "fixed", so just update lspconfig.

Ok. To answer everyone's question.

The built-in client can send multiple different unicode encodings depending on the server. UTF-16 is the most common, but UTF-8 is saner/more efficient for most editors to provide. Clangd provides utf-8 support via the off-spec offsetEncoding option, which was previously set in on_init (on the client) in lspconfig. You all were overriding on_init, which caused the offset encoding to mismatch what clangd was sending.

Lspconfig always wraps your on_init to send workspace/didChangeConfiguration after initialization anyways, so I just added a small check for offsetEncoding in this wrapper to set offsetEncoding if it was found. Now you can go ahead and override on_init, and it should "just work", since the wrapper will handle everything.

You should still read all of the lspconfig hooks before overriding them, because there are times where we have to do workarounds for servers (like julia/omnisharp with on_new_config).

**Comment by Leiyi548** [comment]
@mjlbach Yes, I tested it and the problem was solved. Thank you for your help

**Comment by mjlbach** [comment]
I'd also like to apologize to any users affected by this. I don't encounter these UTF issues often as a predominantly ASCII user, but I've been trying to fix them as they are reported :) Hopefully almost all of these issue will be addressed by the time neovim 0.6 is released next month.

**Comment by pizsd** [comment]
> @kylo252 In China, LvimUpdate failed. How can I update it?

你的LunarVim可以输入中文吗，我的安装完成后无法输入中文了
